nova/utils.py

```diff
@@ -1238,26 +1241,32 @@ def is_numa_aligned_flavor(flavor):
                for prop, value in flavor.extra_specs.items())
 
 
-def get_reserved_memory(flavor):
+def get_reserved_memory_and_cpu(flavor):
     # baremetals don't need reservation as they have their whole host
     if is_baremetal_flavor(flavor):
-        return 0
+        return 0, 0
 
-    # explicit definitions in the flavor take precedence over any heuristic
-    try:
-        key = MEMORY_RESERVABLE_MB_RESOURCE_SPEC_KEY
-        memory_mb = int(flavor.extra_specs[key])
-    except (ValueError, KeyError):
-        pass
-    else:
-        if memory_mb > 0:
-            return min(flavor.memory_mb, memory_mb)
+    def _get_reserved_from_flavor_and_conf(key, flavor_value, full_threshold):
+        # explicit definitions in the flavor take precedence over any heuristic
+        try:
+            value = int(flavor.extra_specs[key])
+        except (ValueError, KeyError):
+            return 0
+        else:
+            if value > 0:
+                return min(value, flavor_value)
+        if 0 <= full_threshold <= flavor_value:
+            return flavor_value
+        return 0
```

Comment:
Would it make sense to pass in `flavor`, too, so that the function doesn't rely on being nested and able to access it from the "outside"?

Why do we `return 0` now instead of `pass`? This leads to us no longer handling `full_threshold` (or `CONF.full_reservation_memory_mb`) if the `key` does not exist. (Maybe write some tests for this method?)

Reply:
passing in flavor would require us to pass in the attr for `flavor_value` as string, which i'm not a fan of (no type checking). the alternative would be to do `flavor.extra_specs[...SPEC_KEY]` on the outside, but then we lose the dry `try ... except KeyError` version.
so yeah, i get your point about the duplicate methods of parameter-passing, but i don't agree that i'd have to take care of being able to call the method from the outside. this method is there purely for dry-ness, not really for re-usability elsewhere.

i fixed the `pass`.

Reply:
added some test cases